### PR TITLE
CDAP-12191 expose pipeline info for each plugin context

### PR DESCRIPTION
**Pull request by albertshau**
Adds methods to get pipeline name, namespace, logical start time,
and output port schemas to StageContext so that every plugin has
access to them at runtime. Also doing some cleanup and refactoring
so that exposing more information in the future will be easier.

Consolidated common pipeline info and CDAP contexts into a
PipelineRuntime class that is passed around in all the plugin
context constructors. The next time we need to expose more fields,
or CDAP capabilities, we can just modify that class instead of
modifying every single constructor in the context hierarchy.
Also removed as many of the constructors as possible to try and
simplify things.

Also removed the getHadoopJob method that has been deprecated since
3.x and which usually just throws an UnsupportedOperationException.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT6013

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 1]
```diff
@@ -175,6 +178,8 @@ public void initialize() throws Exception {
     if (Boolean.valueOf(properties.get(Constants.STAGE_LOGGING_ENABLED))) {
       LogStageInjector.start();
     }
+    PipelineRuntime pipelineRuntime = new PipelineRuntime(context, mrMetrics);
+    Admin admin = context.getAdmin();
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: seems weird to get/assign it 100 lines from its first usage.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageContext.java [hunk 2]
```diff
@@ -121,6 +143,15 @@
   Schema getOutputSchema();
 
   /**
+   * Return the output port schemas for the stage. The map key is the port name and the map value is the schema for
+   * that port. This map is the same map that was set at configure time by {@link MultiOutputPipelineConfigurable}.
+   * Stages that do not implement {@link MultiOutputPipelineConfigurable} should not use this method.
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageContext.java]: whats the behavior when stages that do not implement MultiOutputPipelineConfigurable use this method? does it throw an exception?

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/batch/BasicSparkPluginContext.java [hunk 3]
```diff
@@ -33,29 +32,17 @@
 
   private final SparkClientContext sparkContext;
 
-  public BasicSparkPluginContext(SparkClientContext sparkContext, StageSpec stageSpec) {
-    super(sparkContext, sparkContext.getMetrics(), new DatasetContextLookupProvider(sparkContext),
-          sparkContext.getLogicalStartTime(), sparkContext.getAdmin(), stageSpec, new BasicArguments(sparkContext));
+  public BasicSparkPluginContext(SparkClientContext sparkContext, PipelineRuntime pipelineRuntime, StageSpec stageSpec,
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/batch/BasicSparkPluginContext.java]: mark the first parameter `@Nullable`?

**Comment by sagarkapare** [comment]
One question, otherwise LGTM 👍

**Comment by anwar6953** [comment]
Two comments, otherwise LGTM 👍

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageContext.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageContext.java]: it won't throw an exception, but will return an empty map. Will add that to javadoc

**Comment by albertshau** [comment]
fixed and squashed, will merge after a green build

**Pull request merged**